symbolic/src/Data/Macaw/Symbolic.hs

```diff
@@ -892,6 +909,97 @@ toCoreCFG :: MacawSymbolicArchFunctions arch
           -> C.SomeCFG (MacawExt arch) init ret
 toCoreCFG archFns (CR.SomeCFG cfg) = crucGenArchConstraints archFns $ C.toSSA cfg
 
+-- * Retaining original macaw block endings when slicing CFG
+
+data MacawBlockEndCase =
+    MacawBlockEndJump
+  | MacawBlockEndCall
+  | MacawBlockEndReturn
+  | MacawBlockEndBranch
+  | MacawBlockEndArch
+  | MacawBlockEndFail
+  deriving (Eq, Ord, Enum, Bounded)
+
+data MacawBlockEnd arch = MacawBlockEnd MacawBlockEndCase !(Maybe (M.ArchSegmentOff arch))
+
+-- we encode 'MacawBlockEndCase' as a 3-bit bitvector
+type MacawBlockEndType arch = C.StructType (Ctx.EmptyCtx Ctx.::> C.BVType 3 Ctx.::> C.MaybeType (MM.LLVMPointerType (M.ArchAddrWidth arch)))
```

Comment:
Can we turn this into a haddock comment for nicer rendering, and can we document what the second struct element is?